composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+        self._done_loading()
+
+    def _load(self) -> None:
+        """Load shards."""
+        # We find out num workers, and therefore num partitions, when __iter__ is called.
+        # From the partition, derive our shard overlap range and exact sample range.
+        world = World.from_env()
+        todo_shards, part_min_id, part_max_id = self.index.get_partition(world, self.batch_size)
+
+        # Preload all of our shards that are already available in the cache.
+        todo_shards = self._load_shards_if_downloaded(todo_shards, part_min_id, part_max_id)
+
+        # Start downloading our missing shards in a background thread, if there are any.
+        if todo_shards:
+            thread = Thread(target=self._download_thread, args=(todo_shards, part_min_id, part_max_id), daemon=True)
+            thread.start()
+        else:
+            self._done_loading()
+
+    def __len__(self) -> int:
+        """Get the length of the dataset.
+
+        Returns:
+            int: Dataset length.
+        """
+        return math.ceil(self.index.total_samples / dist.get_world_size())
+
+    def _unpack_sample(self, data: bytes) -> Dict[str, Any]:
+        """Unpack a sample dict from raw bytes.
+
+        First unpacks the str to raw bytes dict, then unpacks each field's raw bytes.
+
+        Args:
+            data (bytes): The packed bytes of the sample.
+
+        Returns:
+            Dict[str, Any]: The sample dict.
+        """
+        key_to_raw = bytes_to_sample_dict(data, self.index.fields)
+        obj = {}
+        for key, decode in self.decoders.items():
+            value = key_to_raw[key]
+            if decode:
+                value = decode(value)
+            obj[key] = value
+        return obj
+
+    def __getitem__(self, idx: int) -> Any:
+        """Get the sample at the index, assuming its shard is loaded.
+
+        Do not call this directly unless all shards have been loaded. Will crash
+        if the shard is not loaded.
+
+        Args:
+            idx (int): Sample ID.
+
+        Returns:
+            Any: The sample.
+        """
+        shard = self.index.sample_shards[idx]
+        offset = self.index.sample_shard_offsets[idx]
+        size = self.index.bytes_per_sample[idx]
+
+        basename = get_shard_basename(shard)
+        shard_filename = os.path.join(self.local, basename)
+        fp = open(shard_filename, 'rb', 0)
+        fp.seek(offset)
+        data = fp.read(size)
+        fp.close()
+
+        return self._unpack_sample(data)
+
+    def _new_growing_epoch(self) -> int:
+        """Start a new growing epoch, in which we own the sample sequence because it grows.
+
+        Returns:
+            int: The epoch ID, an identifier which is given back to the caller.
+        """
+        with self._lock:
+            epoch = self._next_epoch
+            self._next_epoch += 1
+            self._epoch_to_todo_ids[epoch] = list(self._downloaded_ids)
```

Comment:
Do we need to track epoch numbers, or can we reset the current buffer of TODO ids? E.g. at the beginning of each epoch / iteration, we set `self._todo_ids = list(self._download_ids)`. I think it is OK to assume that there is only one active iterator per dataset at any given time (we don't iterate over a dataset while already iterating over it).

Reply:
Will attempt to refactor, would like to add a note though that in the future we want to support fast resumption, which may/may not require an `epoch` field, depending on how we implement it.

Reply:
Tracking this but not addressing in this PR.